File: src/app/model/AltitudeMetric.ts
import { LocationServiceEvent } from './LocationServiceEvent';
import { MetricService } from "../services/metric.service";
import { Metric, MetricType } from "./Metric";

export class AltitudeMetric extends Metric {
    private handler = this.locationEventHandler.bind(this);

    constructor(metricService: MetricService) {
        super(MetricType.ALTITUDE, 'Altitude', 'm', metricService);
    }

    startLogging(): void {
        this.metricService.getLocationService().subscribeForLocation(this.handler);
    }

    stopLogging(): void {
        this.metricService.getLocationService().unsubscribeForLocation(this.handler);
    }

    locationEventHandler(event: LocationServiceEvent): void {
        const altitude = event.location.coords.altitude;
        if (altitude !== null) {
            this.addValue(altitude, new Date(event.location.timestamp));
        }
    }
}

File: android/app/src/main/java/com/openbikecomputer/model/AltitudeMetric.java
package com.openbikecomputer.model;

import com.openbikecomputer.service.IMetricService;
import com.openbikecomputer.service.OnLocationUpdate;

public class AltitudeMetric extends Metric implements OnLocationUpdate {
    public AltitudeMetric(IMetricService metricService) {
        super(MetricType.ALTITUDE, "Altitude", "m", metricService);
    }

    @Override
    public void startLogging() {
        metricService.getLocationService().subscribeForLocation(this);
    }

    @Override
    public void stopLogging() {
        metricService.getLocationService().unsubscribeForLocation(this);
    }

    @Override
    public void onLocation(LocationSample sample) {
        if (sample.altitude != null) {
            addValue(sample.altitude, sample.timestamp);
        }
    }
}
